src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -80925,6 +80925,34 @@ static void F(object? x)
                 Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(23, 19));
         }
 
+        [Fact]
+        [WorkItem(49472, "https://github.com/dotnet/roslyn/issues/49472")]
+        public void TypeInference_13()
+        {
+            var source =
+@"#nullable enable
+using System;
+using System.Collections.Generic;
+static class Program
+{
+    static void Main()
+    {
+        var d1 = new Dictionary<string, string?>();
+        var d2 = d1.ToDictionary(kv => kv.Key, kv => kv.Value);
+        d2[""""].ToString();
+    }
+    static Dictionary<TKey, TValue> ToDictionary<TSource, TKey, TValue>(this IEnumerable<TSource> e, Func<TSource, TKey> k, Func<TSource, TValue> v)
+    {
+        return new Dictionary<TKey, TValue>();
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
```

Comment:
>comp.VerifyDiagnostics( [](start = 12, length = 23)

Would it also make sense to assert nullability information exposed by SemanticModel?